File: scripts/value_semantics/dlist/efficient_val.cc
// Doubly-linked list — Value (index-based)
// Usage: ./efficient_val <mode> <N> [seed]
//   mode: append | traverse | remove_all | remove_half_traverse
// Output: result integer to stdout

#include <climits>
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <string>
#include <vector>

static inline uint64_t lcg_next(uint64_t& state) {
    state = state * 6364136223846793005ULL + 1442695040888963407ULL;
    return state;
}

const int NONE = INT_MAX;

struct Node {
    int pred, succ, value;
};

static long long do_append(int N, uint64_t /*rng*/) {
    std::vector<Node> nodes;
    nodes.reserve(N);
    int head = NONE, tail = NONE;
    for (int i = 0; i < N; ++i) {
        int idx = (int)nodes.size();
        int pred = (tail != NONE) ? tail : NONE;
        nodes.push_back({pred, NONE, i});
        if (tail != NONE) nodes[tail].succ = idx;
        else head = idx;
        tail = idx;
    }
    return 0;
}

static long long do_traverse(int N, uint64_t /*rng*/) {
    std::vector<Node> nodes;
    nodes.reserve(N);
    int head = NONE, tail = NONE;
    for (int i = 0; i < N; ++i) {
        int idx = (int)nodes.size();
        int pred = (tail != NONE) ? tail : NONE;
        nodes.push_back({pred, NONE, i});
        if (tail != NONE) nodes[tail].succ = idx;
        else head = idx;
        tail = idx;
    }
    long long sum = 0;
    for (int cur = head; cur != NONE; cur = nodes[cur].succ)
        sum += nodes[cur].value;
    return sum;
}

static long long do_remove_all(int N, uint64_t rng) {
    std::vector<Node> nodes;
    nodes.reserve(N);
    int head = NONE, tail = NONE;
    std::vector<int> addrs;
    for (int i = 0; i < N; ++i) {
        int idx = (int)nodes.size();
        int pred = (tail != NONE) ? tail : NONE;
        nodes.push_back({pred, NONE, i});
        if (tail != NONE) nodes[tail].succ = idx;
        else head = idx;
        tail = idx;
        addrs.push_back(idx);
    }
    for (int i = N - 1; i > 0; --i) {
        int j = (int)(lcg_next(rng) % (i + 1));
        std::swap(addrs[i], addrs[j]);
    }
    for (int i = 0; i < N; ++i) {
        int a = addrs[i];
        int p = nodes[a].pred, s = nodes[a].succ;
        if (p != NONE) nodes[p].succ = s; else head = s;
        if (s != NONE) nodes[s].pred = p; else tail = p;
    }
    return 0;
}

static long long do_remove_half_traverse(int N, uint64_t rng) {
    std::vector<Node> nodes;
    nodes.reserve(N);
    int head = NONE, tail = NONE;
    std::vector<int> addrs;
    for (int i = 0; i < N; ++i) {
        int idx = (int)nodes.size();
        int pred = (tail != NONE) ? tail : NONE;
        nodes.push_back({pred, NONE, i});
        if (tail != NONE) nodes[tail].succ = idx;
        else head = idx;
        tail = idx;
        addrs.push_back(idx);
    }
    for (int i = N - 1; i > 0; --i) {
        int j = (int)(lcg_next(rng) % (i + 1));
        std::swap(addrs[i], addrs[j]);
    }
    for (int i = 0; i < N / 2; ++i) {
        int a = addrs[i];
        int p = nodes[a].pred, s = nodes[a].succ;
        if (p != NONE) nodes[p].succ = s; else head = s;
        if (s != NONE) nodes[s].pred = p; else tail = p;
    }
    long long sum = 0;
    for (int cur = head; cur != NONE; cur = nodes[cur].succ)
        sum += nodes[cur].value;
    return sum;
}

int main(int argc, char** argv) {
    std::string mode = (argc > 1) ? argv[1] : "all";
    int N = (argc > 2) ? std::atoi(argv[2]) : 500000;
    uint64_t rng = (argc > 3) ? (uint64_t)std::atoll(argv[3]) : 42;

    if (mode == "append")              std::cout << do_append(N, rng) << std::endl;
    else if (mode == "traverse")        std::cout << do_traverse(N, rng) << std::endl;
    else if (mode == "remove_all")      std::cout << do_remove_all(N, rng) << std::endl;
    else if (mode == "remove_half_traverse") std::cout << do_remove_half_traverse(N, rng) << std::endl;
    else {
        long long r = 0;
        r += do_append(N, rng);
        r += do_traverse(N, rng);
        r += do_remove_all(N, rng);
        r += do_remove_half_traverse(N, rng);
        std::cout << r << std::endl;
    }
    return 0;
}

File: scripts/value_semantics/dlist/efficient_val.rs
// Doubly-linked list - Value (index-based, SoA parallel arrays)
// This intentionally keeps the same value/index semantics as efficient_val.cc,
// but uses a different memory layout from its Vec<Node> AoS representation.

use std::env::Alloc;
use std::env::Console;
use std::vec::Vec;

const NONE: i32 = 2147483647;

fn lcg_next(state: u64) -> u64 {
    state * 6364136223846793005 + 1442695040888963407
}

struct DList {
    head: i32,
    preds: Vec<i32>,
    succs: Vec<i32>,
    values: Vec<i32>,
}

fn build_list(N: i32) -> DList can Alloc {
    let mut preds: Vec<i32> = Vec::with_capacity(N as u64);
    let mut succs: Vec<i32> = Vec::with_capacity(N as u64);
    let mut values: Vec<i32> = Vec::with_capacity(N as u64);
    let mut head: i32 = NONE;
    let mut i: i32 = 0;
    while i < N {
        let pred: i32 = if i == 0 { NONE } else { i - 1 };
        let succ: i32 = if i == N - 1 { NONE } else { i + 1 };
        preds.push(pred);
        succs.push(succ);
        values.push(i);
        if i == 0 {
            head = i;
        };
        i = i + 1;
    };
    DList { head: head, preds: preds, succs: succs, values: values }
}

fn do_append(N: i32) -> i32 can Alloc {
    let _dl = build_list(N);
    0
}

fn do_traverse(N: i32) -> u64 can Alloc {
    let dl = build_list(N);
    let mut sum: u64 = 0;
    let mut cur: i32 = dl.head;
    while cur != NONE {
        sum = sum + (dl.values[cur] as u64);
        cur = dl.succs[cur];
    };
    sum
}

fn shuffled_indices(N: i32, seed: u64) -> Vec<i32> can Alloc {
    let mut addrs: Vec<i32> = Vec::with_capacity(N as u64);
    let mut i: i32 = 0;
    while i < N {
        addrs.push(i);
        i = i + 1;
    };

    let mut rng: u64 = seed;
    let mut k: i32 = N - 1;
    while k > 0 {
        rng = lcg_next(rng);
        let j: i32 = (rng as i32) % (k + 1);
        let jj: i32 = if j < 0 { j + (k + 1) } else { j };
        let tmp: i32 = addrs[k];
        addrs[k] = addrs[jj];
        addrs[jj] = tmp;
        k = k - 1;
    };
    addrs
}

fn do_remove_all(N: i32, seed: u64) -> i32 can Alloc {
    let mut dl = build_list(N);
    let addrs = shuffled_indices(N, seed);

    let mut m: i32 = 0;
    while m < N {
        let a: i32 = addrs[m];
        let p: i32 = dl.preds[a];
        let s: i32 = dl.succs[a];
        if p != NONE {
            dl.succs[p] = s;
        } else {
            dl.head = s;
        };
        if s != NONE {
            dl.preds[s] = p;
        } else {
        };
        m = m + 1;
    };
    0
}

fn do_remove_half_traverse(N: i32, seed: u64) -> u64 can Alloc {
    let mut dl = build_list(N);
    let addrs = shuffled_indices(N, seed);

    let half: i32 = N / 2;
    let mut m: i32 = 0;
    while m < half {
        let a: i32 = addrs[m];
        let p: i32 = dl.preds[a];
        let s: i32 = dl.succs[a];
        if p != NONE {
            dl.succs[p] = s;
        } else {
            dl.head = s;
        };
        if s != NONE {
            dl.preds[s] = p;
        } else {
        };
        m = m + 1;
    };

    let mut sum: u64 = 0;
    let mut cur: i32 = dl.head;
    while cur != NONE {
        sum = sum + (dl.values[cur] as u64);
        cur = dl.succs[cur];
    };
    sum
}

fn main() {
    let N: i32 = 500000;
    let seed: u64 = 42;

    let a = handle using Alloc::default() {
        do_append(N)
    };
    handle using Console::default() {
        do Console::println("dlist_append={}", a);
    };

    let t = handle using Alloc::default() {
        do_traverse(N)
    };
    handle using Console::default() {
        do Console::println("dlist_traverse={}", t);
    };

    let r = handle using Alloc::default() {
        do_remove_all(N, seed)
    };
    handle using Console::default() {
        do Console::println("dlist_remove_all={}", r);
    };

    let h = handle using Alloc::default() {
        do_remove_half_traverse(N, seed)
    };
    handle using Console::default() {
        do Console::println("dlist_remove_half_traverse={}", h);
    };
}
